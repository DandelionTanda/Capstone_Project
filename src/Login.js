import { StatusBar } from 'expo-status-bar';
import React, { useState } from 'react';
import { StyleSheet, Text, View, Button} from 'react-native';
import {FloatingLabelInput} from "react-native-floating-label-input";
import AsyncStorage from "@react-native-async-storage/async-storage";


function Login({navigation}) {

  const [username, setUsername] = useState('')
  const [password, setPassword] = useState('')
  const [error, setError] = useState(false)

  function DoLogin(){
     
    fetch(`https://my.tanda.co/api/oauth/token`,{
     method: "POST",
     headers: {
       'Content-Type': 'application/json'
     },
     body: JSON.stringify({
       username:username,
       password:password,
       scope:"me user device platform",
       grant_type:"password"
     })
   })
   .then(res=>res.json())
   .then(data=>{  
      try{     
        console.log(data) 
        if (data.error){
          setError(data.error)
        }   
        else {          
        localStorage.setItem('token', data.access_token)
        localStorage.setItem('tokenType', data.token_type) 
        setError(false)      
        navigation.navigate("Discount") 
        }               
      }
      catch(err)
      {
        console.log(err)
        
      }
    })
    
   
  }

  return (
    <View style={styles.container}>
      <StatusBar style="auto" />
      <View style={styles.header}>
      <Text style={styles.Text1}>TANDA</Text>
      <Text style={styles.Text2}>Discount</Text>
      <Text style={styles.Text3}>Discovery</Text>      
      </View >
      {error?
      <Text style={{fontSize: 20, fontWeight: "bold", color: 'red', textAlign: 'center'}}>
        username or password is incorrect
        </Text>:null}
        <View style={styles.footer}>          
        <Text></Text>
        <Text></Text>
        <View >         
          <FloatingLabelInput 
          label= "Username"
          value={username}
          containerStyles={{
          borderWidth:2,
          borderColor:"white",
          borderRadius: 10,
          width:40,
          height:40,
          }}
          onChangeText={value => setUsername(value)}
        /></View>
        <Text></Text>
        <Text></Text>
        <View >
        <FloatingLabelInput
          label="Password"
          isPassword={true}
          value={password}
          containerStyles={{
          borderWidth:2,
          borderColor:"white",
          borderRadius: 10,
          width:40,
          height:40,
          }}
          onChangeText={value => setPassword(value)}
        />
        </View>
        <Text></Text>
        <Text></Text>
        <View>
        <Button title = "Login" onPress = {DoLogin}/>
        </View>
      </View>
    </View>
  ); 
}

export default Login;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#45B8DB',   
  },
  header:{
    marginTop: 60,
    flex:1, 
  },
  Text1:{
    flex:1,
    position:"relative",
    fontWeight:"bold",
    color:"royalblue",
    left:280,
    fontSize:30,
    fontFamily:"sans-serif",
    fontWeight: "bold"
  },
  Text2:{
    flex:1,
    position:"relative",
    color:"white",
    bottom:40,
    left:20,
    fontSize:50,
    fontFamily:"sans-serif",
    fontWeight: "bold"
  },
  Text3:{
    flex:1,
    position:"relative",
    color:"white",
    bottom:60,
    left:120,
    fontSize:50,
    fontFamily:"sans-serif",
    fontWeight: "bold"
  },
  footer:{
    flex:1,
    color:"white",
  },
  username:{
    padding:10,
    position:"relative",
    bottom:60,
  },
  password:{
    padding:10,
    position:"relative",
    bottom:60,
  },
  button:{
    padding:15,
    position:"relative",
    bottom:60,
  }
});
