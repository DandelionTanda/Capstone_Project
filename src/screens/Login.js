import React, { useState } from 'react';
import { StyleSheet, Text, View, Dimensions, PixelRatio} from 'react-native';
import {Button, Icon, Input} from "react-native-elements";
import { color } from 'react-native-elements/dist/helpers';
import MyButton from '../components/MyButton';

// Retrieve initial screen's width
let screenWidth = Dimensions.get('screen').width;

// Retrieve initial screen's height
let screenHeight = Dimensions.get('screen').height;

function fontSizer(screenHeight){
  if(screenHeight > 550){
    return 40;
  }else { 
    return 35;
  }

};

function textSizer(screenWidth,screenHeight){
  if(screenHeight < 550)
  {
    return screenWidth*0.1
  }
  else{
    return screenWidth*0.15
  }
}

function buttonSizer(screenWidth,screenHeight){
  if(screenWidth < 350 || screenHeight <550)
  {
    return screenWidth*0.3
  }
  else{
    return screenWidth*0.2
  }
}


function buttonPadding(screenWidth, screenHeight){
  if(screenWidth < 350 || screenHeight <550)
  {
    return 5
  }
  else{
    return 15
  }
}

function buttonTop(screenWidth, screenHeight){
  if(screenWidth < 350 || screenHeight <550)
  {
    return screenWidth*0.48
  }
  else{
    return screenWidth*0.45
  }
}

function Login({navigation}) {

  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')
  const [error, setError] = useState('') 
  
  async function DoLogin(){   
    
    try {
      const fetchResult = await fetch(`https://internal-allow-partner-organisation-to-be-switched-to.ms.tanda.co/api/oauth/token`,{
        method: "POST",
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          username:"zouweiran9122@gmail.com",
          //username:"leo727268082@gmail.com",
          password:"123456789",
          scope:"me user device platform organisation",
          grant_type:"password"
        })
      })        
      
      if (!fetchResult.ok) {         
        if (fetchResult.status === 400) {               
          const errorMessage = "Invalid email or password";
          throw Error(errorMessage)        
        }
        else {
          const errorMessage = `An error has occured: ${fetchResult.status}`;   
          throw Error(errorMessage)         
        }                  
      }  
      else {
        setError("")
        const data = await fetchResult.json() 
        localStorage.setItem('token', data.access_token)
        localStorage.setItem('tokenType', data.token_type) 
        navigation.navigate("HomeTabs") 
      }
    } catch (err) {      
      
      setError(err.message)
    }   
  };   

  return (
    <View style={styles.container}>    
      <View style={styles.header}>
      <Text style={styles.Text1}>TANDA</Text>
      <Text style={styles.Text2}>Discount</Text>
      <Text style={styles.Text3}>Discovery</Text>      
      </View>
      <View style={styles.main}>
        <Input style={styles.email}
          placeholder='Email'
          placeholderTextColor="white"
          containerStyle={{width:screenWidth, color:"white"}}
          inputStyle={{color:"white", fontSize:16}}
          label="Email"
          labelStyle={{color:"white"}}
          value={email}
          onChangeText={value => setEmail(value)} 
        />

        <Text></Text>

        <Input style={styles.password}
          placeholderTextColor="white"
          placeholder='Password'
          containerStyle={{width:screenWidth, color:"white"}}
          inputStyle={{color:'white', fontSize:16}}
          label="Password"
          labelStyle={{color:"white"}}
          value={password}
          onChangeText={value => setPassword(value)}
          secureTextEntry={true}
          renderErrorMessage={false}
          errorMessage=""
          errorStyle={{ color: 'red' }}
        />
        {error ? <Text style={styles.error} testID='Error'>{error}</Text> : null}
      </View>     
      <MyButton onPress={DoLogin} title={'Login'} buttonStyle={styles.loginButton} textStyle={styles.buttonText}/> 
      
    </View>

  ); 
}

export default Login;

const styles = StyleSheet.create(
  {
  
  container: {
    flex: 1,
    backgroundColor: '#45B8DB',  
    
  },

  Text1:{
    marginTop: textSizer(screenWidth,screenHeight),
    fontWeight:"bold",
    color:"royalblue",
    left:screenWidth*0.68,
    fontSize:27,
    fontFamily:"FredokaOne-Regular",
  },

  Text2:{
    
    color:"white",
    marginTop: textSizer(screenWidth,screenHeight),
    marginLeft: screenWidth*0.1,
    fontSize:fontSizer(screenHeight),
    fontFamily:"Arial",
    fontWeight: "bold"
  },

  Text3:{
    color:"white",
    marginTop: textSizer(screenWidth,screenHeight),
    marginLeft:screenWidth*0.2,
    textAlign:"center",
    fontSize: fontSizer(screenHeight),
    fontFamily:"Arial",
    fontWeight: "bold"
  },

  main:{
    marginTop:screenHeight*0.04,
    marginLeft: screenWidth*0.12,
    marginRight: screenWidth*0.1,
    height:screenHeight*0.1,
    alignItems:'center',
   
  },

  email:{
    borderBottomColor:'white',
    borderBottomWidth: 1.5 
  },

  password:{
    marginTop:-5,
    borderBottomColor:'white',
    borderBottomWidth: 1.5 
  },

  loginButton:{
    alignItems: 'center',
    justifyContent: 'center',
    marginLeft:buttonSizer(screenWidth,screenHeight),
    marginRight:buttonSizer(screenWidth,screenHeight),
    paddingVertical: buttonPadding(screenWidth, screenHeight),
    borderRadius: 4,  
    backgroundColor: 'white',
    position:"relative",
    top:buttonTop(screenWidth, screenHeight),
  },

  buttonText:{
    fontSize:24,
    fontWeight:'bold',
    color:"#45B8DB"
  },
  error: {
    fontSize:20,
    marginTop: 10,
    color:'red',
  }
});
