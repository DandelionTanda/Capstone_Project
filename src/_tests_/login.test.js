import 'react-native'
import React from 'react'
import Login from "../screens/Login"
import renderer from 'react-test-renderer';
import {fireEvent, render} from '@testing-library/react-native'
import { View, Text, Pressable } from 'react-native';

describe('Login setup', () => {
  it('Should render correctly', () => {
    const { toJSON } = render(
      <Login />
    );
    expect(toJSON()).toMatchSnapshot();
  })

  it('should render an email input tag', () => {
    const { getByPlaceholderText } = render(
      <Login />
    );
    expect(getByPlaceholderText('Email')).toBeTruthy();
  });
  
  it('should render a password input tag', () => {
    const { getByPlaceholderText } = render(
      <Login />
    );
    expect(getByPlaceholderText('Password')).toBeTruthy();
  });

  it('should render a login button', () => {
    const { getByText } = render(
      <Login />
    );
    expect(getByText('Login')).toBeTruthy();
  });
  
  it('the default value for both fields should be empty', () => {
    const { getByPlaceholderText } = render(
      <Login />
    );
    const emailInput = getByPlaceholderText('Email');
    const passwordInput = getByPlaceholderText('Password');
    expect(emailInput.props.value).toBe('');
    expect(passwordInput.props.value).toBe('');
  });

  it('Check email input', () => {
    const { getByPlaceholderText } = render(
      <Login />
    );
    const emailInput = getByPlaceholderText('Email');
    fireEvent.changeText(
      emailInput,
      'zouweiran9122@gmail.com'
    );
    expect(emailInput.props.value).toBe('zouweiran9122@gmail.com');
  })
  
  it('Check password input', () => {
    const { getByPlaceholderText } = render(
      <Login />
    );
    const passwordInput = getByPlaceholderText('Password');
    fireEvent.changeText(
      passwordInput,
      '123456789'
    );
    expect(passwordInput.props.value).toBe('123456789');
  })

  it('on pressing, a press handler function should be triggered on click event',  () => {
    const onPressMock = jest.fn();
    const { getByText } = render(
      <View >       
        <Pressable onPress={onPressMock}>
          <Text>Login</Text>
        </Pressable>      
      </View>
    );
    fireEvent.press(getByText('Login'));
    expect(onPressMock).toHaveBeenCalled();
  });  

})
