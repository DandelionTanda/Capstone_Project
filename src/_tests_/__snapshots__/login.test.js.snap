// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login setup Should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#45B8DB",
      "flex": 1,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "royalblue",
          "fontFamily": "FredokaOne-Regular",
          "fontSize": 27,
          "fontWeight": "bold",
          "left": 510.00000000000006,
          "marginTop": 112.5,
        }
      }
    >
      TANDA
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "Arial",
          "fontSize": 40,
          "fontWeight": "bold",
          "marginLeft": 75,
          "marginTop": 112.5,
        }
      }
    >
      Discount
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "Arial",
          "fontSize": 40,
          "fontWeight": "bold",
          "marginLeft": 150,
          "marginTop": 112.5,
          "textAlign": "center",
        }
      }
    >
      Discovery
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 133.4,
        "marginLeft": 90,
        "marginRight": 75,
        "marginTop": 53.36,
      }
    }
  >
    <View
      style={
        Object {
          "color": "white",
          "paddingHorizontal": 10,
          "width": 750,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Email
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          onChangeText={[Function]}
          placeholder="Email"
          placeholderTextColor="white"
          rejectResponderTermination={true}
          style={
            Object {
              "borderBottomColor": "white",
              "borderBottomWidth": 1.5,
              "color": "white",
              "flex": 1,
              "fontSize": 16,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
    <Text />
    <View
      style={
        Object {
          "color": "white",
          "paddingHorizontal": 10,
          "width": 750,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Password
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          onChangeText={[Function]}
          placeholder="Password"
          placeholderTextColor="white"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Object {
              "borderBottomColor": "white",
              "borderBottomWidth": 1.5,
              "color": "white",
              "flex": 1,
              "fontSize": 16,
              "marginTop": -5,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "red",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 4,
        "justifyContent": "center",
        "marginLeft": 150,
        "marginRight": 150,
        "opacity": 1,
        "paddingVertical": 15,
        "position": "relative",
        "top": 337.5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#45B8DB",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      Login
    </Text>
  </View>
</View>
`;
